src/Compilers/CSharp/Portable/Syntax/Syntax.xml

```diff
@@ -2002,12 +2002,23 @@
     <Kind Name="RecursivePattern" />
     <Field Name="Type" Type="TypeSyntax" Optional="true" />
     <Field Name="PositionalPatternClause" Type="PositionalPatternClauseSyntax" Optional="true" />
+    <Field Name="LengthPatternClause" Type="LengthPatternClauseSyntax" Optional="true" />
     <Field Name="PropertyPatternClause" Type="PropertyPatternClauseSyntax" Optional="true" />
```

Comment:
Name is now irrelevant. Should we stick with it?

Reply:
I think we should add another optional element for the list pattern clause.
Then we can enforce that both are not allowed to be provided together.
That takes care of compatibility concerns.

Something like:
```
<Field Name="PropertyPatternClause" Type="PropertyPatternClauseSyntax" Optional="true" />
<Field Name="ListPatternClause" Type="ListPatternClauseSyntax" Optional="true" />
````

What do you think?

Reply:
Ideally this should've been named "object pattern clause" since "property" is not accurate either.
Having a separate node means we have to somehow disambiguate (probably by the first subpattern).
I also considered a additional SyntaxKind for property pattern clause without renaming but I'm not sure if that's any better.
If renaming is not an option I'm open to both. But as far as the compiler concerns, none of these are necessary.

Reply:
Tagging @dotnet/roslyn-compiler @CyrusNajmabadi for some thoughts on object model for list syntax.

Reply:
I'm with Julien on this.  The shapes for the list-pattern `{ p1, p2, p3 }` is very different from teh property/field pattern: `{ name1: p1, name2: p2, ... }`.  I would prefer two separate very distinct shapes here.

Reply:
> since "property" is not accurate either (

Technically true.  But you can think of this not as meaning "a strict C# property" and more like "An aspect  of the value".  in the latter regard, .net props, fields, and indexers all fit into the idea that this is an 'aspect' of the value.

Reply:
Julien's proposal seems like the way to go to me.

Reply:
Cool. Thanks

As for the name "property pattern", we should stick with that. That comes from the [patterns-v2 spec](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-8.0/patterns.md#property-pattern).

Reply:
@alrz I've been thinking about this further and chatting with @CyrusNajmabadi. I'd missed that sub-patterns already allow an optional name-colon. Sorry about that.

The trade-off seems to be: 
1. introduce a separate list with better names and specialized nodes, 
2. stick with current structure and bad names.

Some factors to consider:
- we need to deal with error recovery (mixing names and missing names): `{ 1, Property: 2 }`
- we should think about the future: `{ [0]: 1 }` and `{ Property1.Property2: 3 }`

Cyrus suggested that this is a similar problem/situation as we have object initializers (`InitializerExpressionSyntax`).
What we could do is keep the existing nodes (even though names are not great) as you did, but add a different kind on `PropertyPatternClauseSyntax` like we do for `InitializerExpressionSyntax`. Then we can have a kind with a good name even if we can't rename the type itself...
We decide which kind to use by looking at the patterns in the list (like we do for parsing object/collection initializers).

What do you think?

Reply:
>add a different kind on PropertyPatternClauseSyntax 

That's the one I tried first but reverted since it wasn't necessary *in* the compiler. I see how it could be useful. 

Done.